modules/sdk-coin-ada/src/ada.ts

```diff
@@ -394,161 +416,233 @@ export class Ada extends BaseCoin {
+   * @param {string} [params.endingScanIndex] - receive address index to end scanning at. default to startingScanIndex + 20 (exclusive).
+   */
+  async recoverConsolidations(params: ConsolidationRecoveryOptions): Promise<AdaTx[] | AdaSweepTxs> {
     const isUnsignedSweep = !params.userKey && !params.backupKey && !params.walletPassphrase;
-    const MPC = await EDDSAMethods.getInitializedMpcInstance();
-    const derivationPathPrefix = params.seed ? getDerivationPath(params.seed) : 'm';
-    const stakeKeyPair = new AdaKeyPair({
-      pub: MPC.deriveUnhardened(bitgoKey, derivationPathPrefix + '/0').slice(0, 64),
-    });
+    const startIdx = params.startingScanIndex || 1;
+    const endIdx = params.endingScanIndex || startIdx + DEFAULT_SCAN_FACTOR;
 
-    for (let i = startIdx; i < numIteration + startIdx; i++) {
-      const currPath = derivationPathPrefix + `/${i}`;
-      const accountId = MPC.deriveUnhardened(bitgoKey, currPath).slice(0, 64);
-      const paymentKeyPair = new AdaKeyPair({ pub: accountId });
-      const senderAddr = Utils.default.createBaseAddressWithStakeAndPaymentKey(
-        stakeKeyPair,
-        paymentKeyPair,
-        addrFormat
+    if (startIdx < 1 || endIdx <= startIdx || endIdx - startIdx > 10 * DEFAULT_SCAN_FACTOR) {
+      throw new Error(
+        `Invalid starting or ending index to scan for addresses. startingScanIndex: ${startIdx}, endingScanIndex: ${endIdx}.`
       );
-      const { balance, utxoSet } = await this.getAddressInfo(senderAddr);
-      if (balance <= 0) {
-        continue;
-      }
+    }
 
-      // first build the unsigned txn
-      const tipAbsSlot = await this.getChainTipInfo();
-      const txBuilder = this.getBuilder().getTransferBuilder();
-      txBuilder.changeAddress(params.recoveryDestination, balance.toString());
-      for (const utxo of utxoSet) {
-        txBuilder.input({ transaction_id: utxo.tx_hash, transaction_index: utxo.tx_index });
-      }
-      // each slot is about 1 second, so this transaction should be valid for
-      // 7 * 86,400 seconds (7 days) after creation
-      txBuilder.ttl(Number(tipAbsSlot.abs_slot) + 7 * 86400);
-      const unsignedTransaction = (await txBuilder.build()) as Transaction;
-
-      let serializedTx = unsignedTransaction.toBroadcastFormat();
-      if (!isUnsignedSweep) {
-        if (!params.userKey) {
-          throw new Error('missing userKey');
-        }
-        if (!params.backupKey) {
-          throw new Error('missing backupKey');
-        }
-        if (!params.walletPassphrase) {
-          throw new Error('missing wallet passphrase');
+    const addressParams = {
+      bitgoKey: params.bitgoKey,
+      index: 0,
+      seed: params.seed,
+    };
+    const { address: baseAddress } = await this.getAdaAddressAndAccountId(addressParams);
+    const consolidationTransactions: any[] = [];
+    for (let i = startIdx; i < endIdx; i++) {
```

Comment:
I think there's some argument made to change this to be inclusive so that like, first run I put in 1, 20, next run 21, 40, etc. I'm fine with it though

Reply:
since tron is exclusive too right now, ill leave it as is. but good to note and follow up on later, easy change.